alibi_detect/cd/base.py

```diff
@@ -0,0 +1,215 @@
+        """
+        super().__init__()
+
+        if p_val is None:
+            logger.warning('No p-value set for the drift threshold. Need to set it to detect data drift.')
+
+        if isinstance(preprocess_kwargs, dict):  # type: ignore
+            if not isinstance(preprocess_fn, Callable):  # type: ignore
+                preprocess_fn = preprocess_drift
+            self.preprocess_fn = partial(
+                preprocess_fn,
+                **preprocess_kwargs
+            )
+            keys = list(preprocess_kwargs.keys())
+        else:
+            self.preprocess_fn, keys = None, []
+
+        # optionally already preprocess reference data
+        self.preprocess_X_ref = preprocess_X_ref
+        if preprocess_X_ref and isinstance(self.preprocess_fn, Callable):  # type: ignore
+            self.X_ref = self.preprocess_fn(X_ref)
+        else:
+            self.X_ref = X_ref
+        self.update_X_ref = update_X_ref
+        self.n = X_ref.shape[0]  # type: ignore
+        self.p_val = p_val
+        self.correction = correction
+
+        if isinstance(input_shape, tuple):
+            self.input_shape = input_shape
+        elif 'max_len' in keys:
+            self.input_shape = (preprocess_kwargs['max_len'],)
+        elif isinstance(X_ref, np.ndarray):
+            self.input_shape = X_ref.shape[1:]
+
+        # compute number of features for the univariate tests
+        if isinstance(n_features, int):
+            self.n_features = n_features
+        elif not isinstance(preprocess_fn, Callable):
+            self.n_features = X_ref.reshape(X_ref.shape[0], -1).shape[-1]
+        elif preprocess_X_ref:  # infer features from preprocessed reference data
+            self.n_features = self.X_ref.reshape(self.X_ref.shape[0], -1).shape[-1]
+        else:  # infer number of features after applying preprocessing step
+            X = self.preprocess_fn(X_ref[0:min(X_ref.shape[0], n_infer)])
+            self.n_features = X.reshape(X.shape[0], -1).shape[-1]
+
+        if correction not in ['bonferroni', 'fdr'] and self.n_features > 1:
+            raise ValueError('Only `bonferroni` and `fdr` are acceptable for multivariate correction.')
+
+        # set metadata
+        self.meta['detector_type'] = 'offline'  # offline refers to fitting the CDF for K-S
+        self.meta['data_type'] = data_type
+
+    def preprocess(self, X: Union[np.ndarray, list]) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Data preprocessing before computing the drift scores.
+
+        Parameters
+        ----------
+        X
+            Batch of instances.
+
+        Returns
+        -------
+        Preprocessed reference data and new instances.
+        """
+        if isinstance(self.preprocess_fn, Callable):  # type: ignore
+            X = self.preprocess_fn(X)
+            X_ref = self.X_ref if self.preprocess_X_ref else self.preprocess_fn(self.X_ref)
+            return X_ref, X
+        else:
+            return self.X_ref, X
```

Comment:
I find it slightly hard to follow this logic. What is the use case for the `preprocess_X_ref` flag in `__init__` being set to `False`? If we could avoid having it as a parameter it would make the code easier to read.

Reply:
The main reason is that some preprocessing methods/steps might require access to all the reference data (eg using mean/var), so I wanted to keep things flexible.

Reply:
I see, maybe worth explaining reasoning in the docs (unless I missed it?), can be a separate PR.

Reply:
Yes, good idea